### PR TITLE
Fix) Recursively call `vec_slice()` when subsetting data frames

### LINKED ISSUE #142: vec_slice() doesn't work correctly with data frame columns
I'm not sure if you were trying to support these yet, but data frame columns (and matrix columns) don't work with `vec_slice()` currently. One solution might be to call `vec_slice()` instead of `[` when you iterate over the columns of the data frame. I don't know if this greatly slows things down or not.

``` r
library(vctrs)
library(tibble)

x <- tibble(x = c(1,2), y = tibble(z = c(2, 5), y = c(3, 4)))
x
#> # A tibble: 2 x 2
#>       x   y$z    $y
#>   <dbl> <dbl> <dbl>
#> 1     1     2     3
#> 2     2     5     4

res <- vec_slice(x, 1L)
res
#> # A tibble: 1 x 2
#>       x   y$z
#>   <dbl> <dbl>
#> 1     1     2

str(res)
#> Classes 'tbl_df', 'tbl' and 'data.frame':    1 obs. of  2 variables:
#>  $ x: num 1
#>  $ y:Classes 'tbl_df', 'tbl' and 'data.frame':   2 obs. of  1 variable:
#>   ..$ z: num  2 5

vec_slice2 <- function (x, i) {
  if (is.logical(i)) {
    i <- which(i)
  }
  stopifnot(is.integer(i) || is.character(i))
  if (is.null(x)) {
    NULL
  }
  else if (is.data.frame(x)) {
    out <- lapply(x, vec_slice, i)
    #out <- lapply(x, `[`, i)
    vec_restore(out, x)
  }
  else if (is_vector(x)) {
    d <- vec_dims(x)
    if (d == 1) {
      if (is.object(x)) {
        x[i]
      }
      else {
        x[i, drop = FALSE]
      }
    }
    else if (d == 2) {
      x[i, , drop = FALSE]
    }
    else {
      miss_args <- rep(list(missing_arg()), d - 1)
      eval_bare(expr(x[i, !!!miss_args, drop = FALSE]))
    }
  }
  else {
    stop("`x` must be a vector", call. = FALSE)
  }
}

vec_slice2(x, 1L)
#> # A tibble: 1 x 2
#>       x   y$z    $y
#>   <dbl> <dbl> <dbl>
#> 1     1     2     3
```

Created on 2018-11-26 by the [reprex
package](http://reprex.tidyverse.org) (v0.2.0).

### PR DESCRIPTION
**Pull request by DavisVaughan**
I also added a test for data frame columns (I believe that is the correct base R approach). Feel free to drop it if you don't think it is necessary.

Closes #142

### CODE REVIEW

File: tests/testthat/test-size.R [hunk 1]
```diff
@@ -49,6 +49,15 @@ test_that("can subset object of any dimensionality", {
   expect_equal(vec_slice(x4, 1L), ones(1, 3, 4, 5))
 })
 
+test_that("can subset data frame columns", {
+  df <- data.frame(x = c(1, 2), y = I(data.frame(z = c(1, 2), w = c(1, 2))))
```
Comment (hadley) [inline tests/testthat/test-size.R]: I think it's slightly better to add after the fact:

```R
df <- data.frame(x = 1:2)
df$y <- data.frame(a = 2:1)
```

File: tests/testthat/test-size.R [hunk 2]
```diff
@@ -49,6 +49,15 @@ test_that("can subset object of any dimensionality", {
...
+
+  base_slice <- df[1, , drop = FALSE]
+  rownames(base_slice) <- NULL
+
+  expect_equal(vec_slice(df, 1L), base_slice)
```
Comment (hadley) [inline tests/testthat/test-size.R]: I think you could be more direct here:

```R
expect_equal(vec_slice(df, 1L)$y, vec_slice(df$y, 1L))
```

File: tests/testthat/test-size.R (hunk 1 above)
Comment (DavisVaughan) [reply tests/testthat/test-size.R]: That makes more sense, especially considering I couldn't even print `df` before

File: tests/testthat/test-size.R (hunk 2 above)
Comment (DavisVaughan) [reply tests/testthat/test-size.R]: 👍

**Comment by hadley** [comment]
Can you resubmit this somehow?

**Pull request merged**